[2.7] Add post and pre upgrade resource validation tests

caliskanugur opened the pull request:
- [x] This PR relies on  #38404 and #38171 and requires a rebase after these are reviewed and merged. 
- [x] This PR relies on #39341. And requires updates for workloads according to the changes.

**Extensions Updates**
- Implemented:
  - A function to access an ingress externally to `extensions/ingresses/ingresses.go`

**Validation Tests**
- Implemented:
  - Post-upgrade and pre-upgrade resource validation cases
  - Private wait healthy response helper built on top of the access an ingresses helper
  - Private wait ingress hostname update helper built on top of the wait package
  - Private string prefix helpers to get an item and check an item in a slice of strings
  - Private name constructor to generate random suffixes with given names

izaac commented on tests/v2/validation/upgrade/upgrade.go:
```diff
@@ -0,0 +1,164 @@
+
+// waitUntilIngressHostnameUpdates is a private function to wait until the ingress hostname updates
+func waitUntilIngressHostnameUpdates(client *rancher.Client, clusterID, namespace, ingressName string) error {
+	timeout := int64(60 * 5)
```
Does this override the default timeout? Assuming 5 minutes (?) does the Update takes this much?

caliskanugur replied:
Yes, this does override the default timeout. 	`WatchTimeoutSeconds = int64(60 * 30)`, the default timeout is 30 minutes, instead of the default one, it uses a 5 minutes timeout. The update usually takes around 1 - 2 mins, making this a bit longer to avoid connectivity issues.

Would it be better if it's around 3.5 minutes?

izaac replied:
I think giving it that extra time would be ok

izaac commented on tests/v2/validation/upgrade/upgrade_test.go:
```diff
@@ -0,0 +1,450 @@
+	}
+
+	u.T().Logf("Running the pre-upgrade checks")
+	u.TestPreUpgrade()
```
It's probably me not understanding but why are we calling the pre-upgrade from the post-upgrade

caliskanugur replied:
The pre-upgrade test creates the resources with randomly generated names and validates the creations and additional steps. Since all the names are randomly generated, this gives the advantage of running those steps again after the post-upgrade checks. 

This step is to call the pre-upgrade test again to rerun the pre-upgrade test steps.

izaac reviewed (changes requested):
Theres a typo that's why I am requesting a code change.

Everything else are just questions to understand the logic better from my end.

caliskanugur commented on tests/framework/extensions/charts/charts.go:
```diff
@@ -56,6 +56,11 @@ type RancherMonitoringOpts struct {
 	RKEScheduler         bool
 }
 
+// RancherMonitoringOpts is a struct of the required options to install Rancher Monitoring with desired chart values.
```
This is [addressed](https://github.com/rancher/rancher/pull/38171/files#diff-0a4ec8a816e091b0440e83bd6ea0e576da568c81bd2a04d7d25082daa2ecd313R59) with #38171 the extensions PR

caliskanugur commented on tests/framework/extensions/charts/rancherlogging.go:
```diff
@@ -0,0 +1,156 @@
+			return err
+		}
+
+		adminClient, err := rancher.NewClient(client.RancherConfig.AdminToken, client.Session)
```
This admin client is instantiated in order to watch the chart whether it's deleted successfully or not. The actual deletion happens beforehand as a standard or admin user depending on the provided key. (L71-77).

caliskanugur commented on tests/v2/validation/upgrade/upgrade_test.go:
```diff
@@ -0,0 +1,450 @@
+	err = charts.WatchAndWaitDeployments(client, u.project.ClusterID, u.namespace.Name, metav1.ListOptions{})
+	require.NoError(u.T(), err)
+
+	u.T().Log("Creating a daemonset with the test container")
```
Great catch, this is not exactly what you're asking for, but this can be addressed and enhanced with this [function](https://github.com/rancher/rancher/blob/release/v2.6/tests/framework/extensions/charts/charts.go#L202-L204).

We need a function that distinguishes workers to achieve what you're asking for. And on top of that, we would need a helper between that list of workers and daemon sets statuses. @igomez06 would this be the correct way to do it?

igomez06 replied:
Yes, if you use the client.Management.Nodes api call to get node info, it should tell you if the node is a worker node or not, and then you can compare the number of workers with daemonset replicas.

caliskanugur replied:
Updated: Added watch and wait for daemon sets, and as the next step added validation of daemon set replicas are equal to the number of workers in the cluster

caliskanugur commented on tests/framework/extensions/ingresses/ingresses.go:
```diff
@@ -0,0 +1,50 @@
+// AccessIngressExternally checks if the ingress is accessible externally,
+// it returns true if the ingress is accessible, false if it is not, and an error if there is an error.
+func AccessIngressExternally(client *rancher.Client, hostname string, isWithTLS bool) (bool, error) {
+	result, err := charts.GetChartCaseEndpoint(client, hostname, "", isWithTLS)
```
This is resolved.

igomez06 commented on tests/v2/validation/upgrade/upgrade_test.go:
```diff
@@ -0,0 +1,450 @@
+	require.NoError(u.T(), err)
+	assert.Equal(u.T(), createdSecret.Name, u.names.random["secretName"])
+
+	podTemplateWithSecretVolume := corev1.PodTemplateSpec{
```
In all these resource instantiations  Ugur, I think it would be best to write constructors for them, and put them in the respective extensions packages. Like the PodTemplateSpec, IngressSpec, etc. What do you think? These object setups are quite large, so I think was could benefit for having them written as constructors.

caliskanugur replied:
Updated

igomez06 reviewed (commented):
changes look good. Just want the Steve PR merged first before anything else goes into the framework

izaac commented on tests/v2/validation/upgrade/workload_test.go:
```diff
@@ -0,0 +1,358 @@
+	assert.True(u.T(), doesNamespaceExist)
+
+	if !doesNamespaceExist {
+		u.T().Skipf("Namespace with prefix %s doesn't exists", u.names.core["namespaceName"])
```
nit pick

In the kubernetes upgrade version PR this is `doesn't exist` and here `doesn't exists`.

I think, I can be wrong though that it is `doesn't exist`.

caliskanugur replied:
Thank you, this should be updated as well.

izaac reviewed (commented):
PR needs conflicts resolved

caliskanugur commented:
Updated the v1 steve client changes, ready to review.

Additionally, moved templates from Kube API extensions to rancher extensions, and added logging chart CRD uninstall.

igomez06 reviewed (commented):
Looks good Ugur! Same as the others would like to see some freeform jobs

caliskanugur commented:
Backport: #39570

igomez06 commented on tests/framework/extensions/kubeapi/workloads/template.go:
```diff
@@ -11,16 +11,6 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// NewContainer is a contructor that creates a container for a pod template i.e. corev1.PodTemplateSpec
```
What's the point of this file?

caliskanugur replied:
I've moved everything except this `NewImagePullSecret`, an extension that uses the dynamic client to create an image pull secret for a pod template and the necessary conversions.

igomez06 reviewed (commented):
looks good, just one question about one of the template files, then it'll get approved

igomez06 reviewed (approved):
Looks good!

igomez06 reviewed (approved):
Looks Good

izaac reviewed (approved):
LGTM
